### DISCUSSION TITLE
Best way to allow user to adjust defaults?

### DISCUSSION THREAD

**Discussion by gelvidge** [maxGraph/maxGraph#83] [Development]
Hi all, Been following the fantastic work that has been done on this project and thought I would try to help out.  I've been working on fixing the examples and have most of them working (I can submit a PR).  A remaining issue I'm having is how best to give the option for users to adjust some of the defaults.  As an example, in VertexHandler class we have 'rotationEnabled' to allow the user to specify whether a rotation handler is enabled.  Previously this was set on the prototype and thus could be adjusted by user for all instances by modifying the prototype.  Currently(ES6), the field is an instance field and this field is used in the constructor of VertexHandler meaning that when the user subclasses the VertexHandler in the Handles.stories example and tries to adjust the default, the update has no effect (because the parent constructor has already been called and used the default value).

So we could:
1)Revert back to saving the user editable defaults on the prototype and expect user to modify the prototype
2)Apply a separate init() method that contains the constructor code so the user can call init separately (and after the defaults have been changed) - potentially can cause downstream impacts
3)Some other methods to allow user to control defaults?

Interesting in hearing others thoughts

**Comment by tbouffard** [comment]
Hello @gelvidge and sorry for the late answer
It's nice to hear that you are willing to contribute. About fixing the examples, if some elements you plan to fix are already listed in https://github.com/maxGraph/maxGraph/issues/73, feel free to add a comment.

**Regarding the general question about default values**
I don't have a clear answer: today, the code has been migrated from mxGraph and AFAIK, no discussion nor implementation started to change the way defaults can be changed.
I remember than I mention this few months ago in https://github.com/maxGraph/maxGraph/pull/77#pullrequestreview-861815060, but it doesn't seem to be closed to what you are looking for.


**But I have some insights about the `VertexHandler` class used in the `Handles.stories` example**
In the following, I refer to the latest available commit of the development branch at the time of writing.

I am surprised that you say "parent constructor has already been called and used the default value". AFAIK, `rotationEnabled` is not called or set in constructor but elsewhere.
https://github.com/maxGraph/maxGraph/blob/443ebf7a520b7e2a222c03204f9d5bde17fdcc8e/packages/core/src/view/handler/VertexHandler.ts#L332-L335
So setting the `rotationEnabled` value after the `VertexHandler`  object instantiation should work.

Having a look at `Handles.stories.js`, I see that the `rotationEnabled` property is defined again in `MyCustomVertexHandler` and this is probably the issue. This is not the property that is used in the `VertexHandler` parent class, so the new value is not used in the `VertexHandler` methods.
https://github.com/maxGraph/maxGraph/blob/443ebf7a520b7e2a222c03204f9d5bde17fdcc8e/packages/html/stories/Handles.stories.js#L81-L85



**_Suggestion_**
Remove the values set in `MyCustomVertexHandler` and set the values when instantiating the new handler 
- instead of
https://github.com/maxGraph/maxGraph/blob/443ebf7a520b7e2a222c03204f9d5bde17fdcc8e/packages/html/stories/Handles.stories.js#L191-L195
- do
```javascript
  class MyCustomGraph extends Graph {
    createVertexHandler(state) {
      const handler = new MyCustomVertexHandler(state);
      handler.livePreview = true;
      handler.rotationEnabled = true;
      return handler;
    }
  }
```

Alternatively, you could set the values in the `MyCustomVertexHandler` constructor.

It should also be possible to continue using prototype assignment as it was previously done with mxGraph.
We already have `VertexHandler.prototype.rotationEnabled = true;` in `Stencils.stories.js`. 

https://github.com/maxGraph/maxGraph/blob/443ebf7a520b7e2a222c03204f9d5bde17fdcc8e/packages/html/stories/Stencils.stories.js#L65-L66

**Reply by gelvidge** [reply]
Thanks @tbouffard - yes will add some comments to the other thread regarding examples.

Regarding the rotationEnabled variable in the Handles example,  as I understand it is used in the parent (VertexHandler) constructor through the isRotationHandleVisible() method which is called in the constructor so I don't think the suggested solutions will work in this case (I've tried them).

**Reply by tbouffard** [reply]
My bad, you are right: I didn't check the constructor code carefully. 😄

**Comment by tbouffard** [comment]
Discussions continue in #267 about `VertexHandler.rotationEnabled`
